File: test/cpp_ffi/ffi_channel_test.cc
// FfiChannel Test Suite
//
// Tests the generated FfiChannel implementation that allows using standard
// gRPC clients over FFI transport in C++.
//
// To compile and run:
//   g++ -std=c++17 -I/path/to/protobuf/include ffi_channel_test.cc -o ffi_channel_test
//   ./ffi_channel_test

#include <cassert>
#include <cstring>
#include <iostream>
#include <map>
#include <memory>
#include <string>
#include <vector>

// =============================================================================
// Mock Protobuf Messages (simplified for testing without full protobuf)
// =============================================================================

class MockMessage {
public:
    virtual ~MockMessage() = default;
    virtual std::string SerializeAsString() const = 0;
    virtual bool ParseFromString(const std::string& data) = 0;
};

class Empty : public MockMessage {
public:
    std::string SerializeAsString() const override { return ""; }
    bool ParseFromString(const std::string& data) override { return true; }
};

class PingResponse : public MockMessage {
public:
    std::string message;

    std::string SerializeAsString() const override {
        return "ping:" + message;
    }

    bool ParseFromString(const std::string& data) override {
        if (data.find("ping:") == 0) {
            message = data.substr(5);
            return true;
        }
        return false;
    }
};

class GetCacheRequest : public MockMessage {
public:
    std::string store_name;
    std::string key;

    std::string SerializeAsString() const override {
        return "get:" + store_name + ":" + key;
    }

    bool ParseFromString(const std::string& data) override {
        if (data.find("get:") == 0) {
            auto rest = data.substr(4);
            auto pos = rest.find(':');
            if (pos != std::string::npos) {
                store_name = rest.substr(0, pos);
                key = rest.substr(pos + 1);
                return true;
            }
        }
        return false;
    }
};

class GetCacheResponse : public MockMessage {
public:
    std::string value;

    std::string SerializeAsString() const override {
        return "resp:" + value;
    }

    bool ParseFromString(const std::string& data) override {
        if (data.find("resp:") == 0) {
            value = data.substr(5);
            return true;
        }
        return false;
    }
};

// =============================================================================
// Mock FfiServer Implementation
// =============================================================================

class MockFfiServer {
public:
    int ping_count = 0;
    int get_count = 0;
    std::string last_key;
    std::map<std::string, std::string> cache_data;

    PingResponse Ping(const Empty& request) {
        ping_count++;
        PingResponse resp;
        resp.message = "pong";
        return resp;
    }

    GetCacheResponse Get(const GetCacheRequest& request) {
        get_count++;
        last_key = request.key;
        GetCacheResponse resp;
        auto it = cache_data.find(request.key);
        if (it != cache_data.end()) {
            resp.value = it->second;
        }
        return resp;
    }
};

// =============================================================================
// Mock Invoke Function (simulates generated Invoke)
// =============================================================================

std::string Invoke(MockFfiServer* server, const std::string& method, const std::string& data) {
    if (method == "/core.v1.HealthService/Ping") {
        Empty req;
        req.ParseFromString(data);
        PingResponse resp = server->Ping(req);
        return resp.SerializeAsString();
    }
    if (method == "/core.v1.CacheService/Get") {
        GetCacheRequest req;
        req.ParseFromString(data);
        GetCacheResponse resp = server->Get(req);
        return resp.SerializeAsString();
    }
    return "";  // Unknown method
}

// =============================================================================
// Mock FfiChannel (simulates generated FfiChannel)
// =============================================================================

class MockFfiChannel {
public:
    explicit MockFfiChannel(MockFfiServer* server) : server_(server) {}

    template<typename Request, typename Response>
    bool InvokeMethod(const std::string& method, const Request& request, Response* response) {
        std::string data = request.SerializeAsString();
        std::string result = Invoke(server_, method, data);
        if (result.empty()) {
            return false;
        }
        return response->ParseFromString(result);
    }

private:
    MockFfiServer* server_;
};

// =============================================================================
// Mock Typed Client (simulates generated HealthServiceFfiClient)
// =============================================================================

class HealthServiceFfiClient {
public:
    explicit HealthServiceFfiClient(MockFfiChannel* channel) : channel_(channel) {}

    bool Ping(const Empty& request, PingResponse* response) {
        return channel_->InvokeMethod("/core.v1.HealthService/Ping", request, response);
    }

private:
    MockFfiChannel* channel_;
};

class CacheServiceFfiClient {
public:
    explicit CacheServiceFfiClient(MockFfiChannel* channel) : channel_(channel) {}

    bool Get(const GetCacheRequest& request, GetCacheResponse* response) {
        return channel_->InvokeMethod("/core.v1.CacheService/Get", request, response);
    }

private:
    MockFfiChannel* channel_;
};

// =============================================================================
// Test Functions
// =============================================================================

void test_ping() {
    std::cout << "test_ping... ";

    MockFfiServer server;
    MockFfiChannel channel(&server);
    HealthServiceFfiClient client(&channel);

    Empty req;
    PingResponse resp;
    bool ok = client.Ping(req, &resp);

    assert(ok);
    assert(resp.message == "pong");
    assert(server.ping_count == 1);

    std::cout << "PASSED" << std::endl;
}

void test_get_cache() {
    std::cout << "test_get_cache... ";

    MockFfiServer server;
    server.cache_data["test-key"] = "test-value";

    MockFfiChannel channel(&server);
    CacheServiceFfiClient client(&channel);

    GetCacheRequest req;
    req.store_name = "default";
    req.key = "test-key";

    GetCacheResponse resp;
    bool ok = client.Get(req, &resp);

    assert(ok);
    assert(resp.value == "test-value");
    assert(server.last_key == "test-key");
    assert(server.get_count == 1);

    std::cout << "PASSED" << std::endl;
}

void test_get_cache_not_found() {
    std::cout << "test_get_cache_not_found... ";

    MockFfiServer server;
    MockFfiChannel channel(&server);
    CacheServiceFfiClient client(&channel);

    GetCacheRequest req;
    req.store_name = "default";
    req.key = "non-existent-key";

    GetCacheResponse resp;
    bool ok = client.Get(req, &resp);

    assert(ok);
    assert(resp.value.empty());

    std::cout << "PASSED" << std::endl;
}

void test_multiple_pings() {
    std::cout << "test_multiple_pings... ";

    MockFfiServer server;
    MockFfiChannel channel(&server);
    HealthServiceFfiClient client(&channel);

    for (int i = 0; i < 100; i++) {
        Empty req;
        PingResponse resp;
        bool ok = client.Ping(req, &resp);
        assert(ok);
    }

    assert(server.ping_count == 100);

    std::cout << "PASSED" << std::endl;
}

void test_channel_with_direct_server() {
    std::cout << "test_channel_with_direct_server... ";

    // Test creating client directly from server (convenience constructor)
    MockFfiServer server;
    MockFfiChannel channel(&server);

    Empty req;
    PingResponse resp;
    bool ok = channel.InvokeMethod("/core.v1.HealthService/Ping", req, &resp);

    assert(ok);
    assert(resp.message == "pong");

    std::cout << "PASSED" << std::endl;
}

// =============================================================================
// Main
// =============================================================================

int main() {
    std::cout << "=== FfiChannel C++ Tests ===" << std::endl;

    test_ping();
    test_get_cache();
    test_get_cache_not_found();
    test_multiple_pings();
    test_channel_with_direct_server();

    std::cout << "=== All tests passed! ===" << std::endl;
    return 0;
}

File: test/rust_ffi/src/ffi_channel_test.rs
// FfiChannel Test Suite
//
// Tests the generated FfiChannel implementation that allows using standard
// gRPC clients over FFI transport in Rust.
//
// To run: cargo test

use std::collections::HashMap;
use std::sync::Arc;

// =============================================================================
// Mock Protobuf Messages (simplified for testing without full prost)
// =============================================================================

pub trait Message: Default {
    fn encode(&self, buf: &mut Vec<u8>) -> Result<(), String>;
    fn decode(buf: &[u8]) -> Result<Self, String>
    where
        Self: Sized;
}

#[derive(Default, Clone)]
pub struct Empty;

impl Message for Empty {
    fn encode(&self, _buf: &mut Vec<u8>) -> Result<(), String> {
        Ok(())
    }

    fn decode(_buf: &[u8]) -> Result<Self, String> {
        Ok(Empty)
    }
}

#[derive(Default, Clone, Debug)]
pub struct PingResponse {
    pub message: String,
}

impl Message for PingResponse {
    fn encode(&self, buf: &mut Vec<u8>) -> Result<(), String> {
        buf.extend_from_slice(format!("ping:{}", self.message).as_bytes());
        Ok(())
    }

    fn decode(buf: &[u8]) -> Result<Self, String> {
        let s = String::from_utf8_lossy(buf);
        if let Some(msg) = s.strip_prefix("ping:") {
            Ok(PingResponse {
                message: msg.to_string(),
            })
        } else {
            Err("Invalid format".to_string())
        }
    }
}

#[derive(Default, Clone)]
pub struct GetCacheRequest {
    pub store_name: String,
    pub key: String,
}

impl Message for GetCacheRequest {
    fn encode(&self, buf: &mut Vec<u8>) -> Result<(), String> {
        buf.extend_from_slice(format!("get:{}:{}", self.store_name, self.key).as_bytes());
        Ok(())
    }

    fn decode(buf: &[u8]) -> Result<Self, String> {
        let s = String::from_utf8_lossy(buf);
        if let Some(rest) = s.strip_prefix("get:") {
            let parts: Vec<&str> = rest.splitn(2, ':').collect();
            if parts.len() == 2 {
                return Ok(GetCacheRequest {
                    store_name: parts[0].to_string(),
                    key: parts[1].to_string(),
                });
            }
        }
        Err("Invalid format".to_string())
    }
}

#[derive(Default, Clone)]
pub struct GetCacheResponse {
    pub value: String,
}

impl Message for GetCacheResponse {
    fn encode(&self, buf: &mut Vec<u8>) -> Result<(), String> {
        buf.extend_from_slice(format!("resp:{}", self.value).as_bytes());
        Ok(())
    }

    fn decode(buf: &[u8]) -> Result<Self, String> {
        let s = String::from_utf8_lossy(buf);
        if let Some(val) = s.strip_prefix("resp:") {
            Ok(GetCacheResponse {
                value: val.to_string(),
            })
        } else {
            Err("Invalid format".to_string())
        }
    }
}

// =============================================================================
// Mock FfiServer Trait and Implementation
// =============================================================================

pub trait FfiServer: Send + Sync {
    fn ping(&self, request: Empty) -> Result<PingResponse, String>;
    fn get(&self, request: GetCacheRequest) -> Result<GetCacheResponse, String>;
}

pub struct MockFfiServer {
    pub ping_count: std::sync::atomic::AtomicUsize,
    pub get_count: std::sync::atomic::AtomicUsize,
    pub cache_data: std::sync::RwLock<HashMap<String, String>>,
}

impl MockFfiServer {
    pub fn new() -> Self {
        MockFfiServer {
            ping_count: std::sync::atomic::AtomicUsize::new(0),
            get_count: std::sync::atomic::AtomicUsize::new(0),
            cache_data: std::sync::RwLock::new(HashMap::new()),
        }
    }

    pub fn set_cache(&self, key: &str, value: &str) {
        self.cache_data
            .write()
            .unwrap()
            .insert(key.to_string(), value.to_string());
    }
}

impl FfiServer for MockFfiServer {
    fn ping(&self, _request: Empty) -> Result<PingResponse, String> {
        self.ping_count
            .fetch_add(1, std::sync::atomic::Ordering::SeqCst);
        Ok(PingResponse {
            message: "pong".to_string(),
        })
    }

    fn get(&self, request: GetCacheRequest) -> Result<GetCacheResponse, String> {
        self.get_count
            .fetch_add(1, std::sync::atomic::Ordering::SeqCst);
        let cache = self.cache_data.read().unwrap();
        let value = cache.get(&request.key).cloned().unwrap_or_default();
        Ok(GetCacheResponse { value })
    }
}

// =============================================================================
// Mock Invoke Function (simulates generated invoke)
// =============================================================================

pub fn invoke<S: FfiServer>(server: &S, method: &str, data: &[u8]) -> Result<Vec<u8>, String> {
    match method {
        "/core.v1.HealthService/Ping" => {
            let request = Empty::decode(data)?;
            let response = server.ping(request)?;
            let mut buf = Vec::new();
            response.encode(&mut buf)?;
            Ok(buf)
        }
        "/core.v1.CacheService/Get" => {
            let request = GetCacheRequest::decode(data)?;
            let response = server.get(request)?;
            let mut buf = Vec::new();
            response.encode(&mut buf)?;
            Ok(buf)
        }
        _ => Err(format!("unknown method: {}", method)),
    }
}

// =============================================================================
// Mock FfiChannel (simulates generated FfiChannel)
// =============================================================================

pub struct FfiChannel<S: FfiServer> {
    server: Arc<S>,
}

impl<S: FfiServer> FfiChannel<S> {
    pub fn new(server: Arc<S>) -> Self {
        FfiChannel { server }
    }

    pub fn invoke<Req: Message, Resp: Message>(
        &self,
        method: &str,
        request: &Req,
    ) -> Result<Resp, String> {
        let mut data = Vec::new();
        request.encode(&mut data)?;
        let result = invoke(&*self.server, method, &data)?;
        Resp::decode(&result)
    }
}

// =============================================================================
// Mock Typed Clients (simulates generated clients)
// =============================================================================

pub struct HealthServiceFfiClient<S: FfiServer> {
    channel: FfiChannel<S>,
}

impl<S: FfiServer> HealthServiceFfiClient<S> {
    pub fn new(server: Arc<S>) -> Self {
        HealthServiceFfiClient {
            channel: FfiChannel::new(server),
        }
    }

    pub fn ping(&self, request: &Empty) -> Result<PingResponse, String> {
        self.channel.invoke("/core.v1.HealthService/Ping", request)
    }
}

pub struct CacheServiceFfiClient<S: FfiServer> {
    channel: FfiChannel<S>,
}

impl<S: FfiServer> CacheServiceFfiClient<S> {
    pub fn new(server: Arc<S>) -> Self {
        CacheServiceFfiClient {
            channel: FfiChannel::new(server),
        }
    }

    pub fn get(&self, request: &GetCacheRequest) -> Result<GetCacheResponse, String> {
        self.channel.invoke("/core.v1.CacheService/Get", request)
    }
}

// =============================================================================
// Tests
// =============================================================================

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_ping() {
        let server = Arc::new(MockFfiServer::new());
        let client = HealthServiceFfiClient::new(server.clone());

        let response = client.ping(&Empty).unwrap();

        assert_eq!(response.message, "pong");
        assert_eq!(
            server.ping_count.load(std::sync::atomic::Ordering::SeqCst),
            1
        );
    }

    #[test]
    fn test_get_cache() {
        let server = Arc::new(MockFfiServer::new());
        server.set_cache("test-key", "test-value");

        let client = CacheServiceFfiClient::new(server.clone());
        let request = GetCacheRequest {
            store_name: "default".to_string(),
            key: "test-key".to_string(),
        };

        let response = client.get(&request).unwrap();

        assert_eq!(response.value, "test-value");
        assert_eq!(
            server.get_count.load(std::sync::atomic::Ordering::SeqCst),
            1
        );
    }

    #[test]
    fn test_get_cache_not_found() {
        let server = Arc::new(MockFfiServer::new());
        let client = CacheServiceFfiClient::new(server.clone());

        let request = GetCacheRequest {
            store_name: "default".to_string(),
            key: "non-existent".to_string(),
        };

        let response = client.get(&request).unwrap();
        assert!(response.value.is_empty());
    }

    #[test]
    fn test_multiple_pings() {
        let server = Arc::new(MockFfiServer::new());
        let client = HealthServiceFfiClient::new(server.clone());

        for _ in 0..100 {
            let response = client.ping(&Empty).unwrap();
            assert_eq!(response.message, "pong");
        }

        assert_eq!(
            server.ping_count.load(std::sync::atomic::Ordering::SeqCst),
            100
        );
    }

    #[test]
    fn test_concurrent_access() {
        use std::thread;

        let server = Arc::new(MockFfiServer::new());
        let mut handles = vec![];

        for _ in 0..10 {
            let server_clone = server.clone();
            let handle = thread::spawn(move || {
                let client = HealthServiceFfiClient::new(server_clone);
                for _ in 0..10 {
                    let _ = client.ping(&Empty).unwrap();
                }
            });
            handles.push(handle);
        }

        for handle in handles {
            handle.join().unwrap();
        }

        assert_eq!(
            server.ping_count.load(std::sync::atomic::Ordering::SeqCst),
            100
        );
    }

    #[test]
    fn test_channel_direct_invoke() {
        let server = Arc::new(MockFfiServer::new());
        let channel = FfiChannel::new(server.clone());

        let response: PingResponse = channel
            .invoke("/core.v1.HealthService/Ping", &Empty)
            .unwrap();

        assert_eq!(response.message, "pong");
    }

    #[test]
    fn test_unknown_method() {
        let server = Arc::new(MockFfiServer::new());
        let channel = FfiChannel::new(server);

        let result: Result<PingResponse, String> =
            channel.invoke("/unknown.Service/Method", &Empty);

        assert!(result.is_err());
        assert!(result.unwrap_err().contains("unknown method"));
    }
}
